randovania/game_description/schema_migration.py

```diff
@@ -69,19 +73,147 @@ def _migrate_v1(data: dict) -> dict:
 
     return data
 
+def _migrate_v2(data: dict) -> dict:
+    game = RandovaniaGame(data["game"])
+    db = data["resource_database"]
+
+    def find_resource(res_type: Union[ResourceTypeLatest, str], index: Optional[int]):
+        if index is None:
+            return None
+        if isinstance(res_type, str):
+            res_type = ResourceTypeLatest(res_type)
+        return migration_data.get_resource_name_from_index(game, index, res_type)
+
+    special_indices = {"energy_tank_item_index", "item_percentage_index", "multiworld_magic_item_index"}
+    for name in special_indices:
+        index = db[name]
+        db[name] = find_resource(ResourceTypeLatest.ITEM, index)
+
+    def migrate_reduction(reduction: dict) -> dict:
+        reduction["name"] = find_resource(ResourceTypeLatest.DAMAGE, reduction.pop("index"))
+        for red in reduction["reductions"]:
+            red["name"] = find_resource(ResourceTypeLatest.ITEM, red.pop("index"))
+        return reduction
+    
+    db["damage_reductions"] = [migrate_reduction(red) for red in db["damage_reductions"]]
+
+    def migrate_requirement(requirement: dict) -> dict:
+        data = requirement["data"]
+
+        if requirement["type"] == "resource":
+            data["type"] = ResourceTypeLatest.from_index(data["type"]).value
```

Comment:
A best practice to the migration functions is to not use project functionality which can then be changed and would break the migration in unexpected ways.

But I think that for the database migration, we can keep this around just for a short while and remove it so it doesn't really matter.